internal/sql/pg/system.go

```diff
@@ -198,6 +202,452 @@ var settingValidations = map[string]settingValidFn{
+	switch ti := tx.(type) {
+	case sql.QueryScanner:
+		return ti.QueryScanFn(ctx, stmt, scans, fn, args...)
+	case conner:
+		conn := ti.Conn()
+		return queryRowFunc(ctx, conn, stmt, scans, fn, args...)
+	}
+	return errors.New("cannot query with scan values")
+}
+
+// FieldDesc describes result value from a query. This is used to convey
+// information on the values passed to the closure given to QueryRowFuncAny.
+type FieldDesc struct {
+	Name                 string
+	TableOID             uint32
+	TableAttributeNumber uint16
+	DataTypeOID          uint32
+	DataTypeSize         int16
+	TypeModifier         int32
+	Format               int16
+}
+
+// QueryRowFuncAny is similar to QueryRowFunc, except that no scan values slice
+// is provided. The provided function is called for each row of the result. The
+// caller does not determine the types of the Go variables in the values slice.
+// In this way it behaves similar to Execute, but providing "for each row"
+// functionality so that every row does not need to be loaded into memory. See
+// also QueryRowFunc, which allows the caller to provide a scan values slice.
+func QueryRowFuncAny(ctx context.Context, tx sql.Executor, sql string,
+	fn func([]FieldDesc, []any) error, args ...any) error {
+	conner, ok := tx.(conner)
+	if !ok {
+		return errors.New("no conn access")
+	}
+	conn := conner.Conn()
+	return queryRowFuncAny(ctx, conn, sql, fn, args...)
+}
+
+func queryRowFuncAny(ctx context.Context, conn *pgx.Conn, sql string,
+	fn func(fields []FieldDesc, vals []any) error, args ...any) error {
+	rows, _ := conn.Query(ctx, sql, args...)
+	fields := rows.FieldDescriptions()
+	pgFields := make([]FieldDesc, len(fields))
+	for i, f := range fields {
+		pgFields[i] = FieldDesc(f)
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		vals, err := rows.Values()
+		if err != nil {
+			return err
+		}
+
+		err = fn(pgFields, vals)
+		if err != nil {
+			return err
+		}
+	}
+
+	return rows.Err()
+}
+
+// TextValue recognizes types used to return SQL TEXT values from SQL queries.
+// Depending on the type and value, the value may represent a NULL as indicated
+// by the null return.
+func TextValue(val any) (txt string, null bool, ok bool) {
+	switch str := val.(type) {
+	case string:
+		return str, false, true
+	case *string:
+		if str == nil { // NULL
+			return "", true, true
+		}
+		return *str, false, true
+	case pgtype.Text: // pgtype.Text uses the Valid field for NULL values
+		return str.String, !str.Valid, true
+	case *pgtype.Text:
+		return str.String, !str.Valid, true
+	}
+	return "", false, false
+}
+
+// ColInfo is used when ingesting column descriptions from PostgreSQL, such as
+// from information_schema.column. Use the Type method to return a canonical
+// ColType. Use the ScanVal method to return a pointer to an instance of the
+// appropriate Go type to scan a row containing this column type in an SQL
+// statement.
+type ColInfo struct {
+	Pos      int
```

Comment:
As mentioned in another comment, I think we should do one of two things:

1. Make the data type uniform with Kwil data types (*types.DataType), and make the `Default` a Kwil type as well (instead of a pgx type), since this is what other functions in this package return.
2. Make this struct unexported, in which case it doesn't matter that it doesn't follow the same semantics as the rest of the package.

Reply:
We could also just get rid of `Default`, if we aren't relying on it

Reply:
Commented in https://github.com/kwilteam/kwil-db/pull/901/files/a62d42727c0574985c662cd8ae94390810691ea5#r1699049173, but regarding `ColInfo`, as the doc comment says, this is really `ingesting column descriptions from PostgreSQL`, and the `Type()` method is intended to return the canonical enumerator of the type.  This is fine as it's well defined and just an enum, but things do get problematic with the `ScanVal() any` method as there's really nothing useful a consumer (outside the `pg` package) can do with the value which is just a type that other functions in the `pg` package want/need.  (You're :100: correct to note that most of this table inspection code is designed to power other functionality internal to `pg`.)